projects/npm-tools/packages/npm-scripts/src/jsp/checkExposesConfiguration.js

```diff
@@ -0,0 +1,93 @@
+/**
+ * SPDX-FileCopyrightText: © 2019 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/* eslint-disable no-console */
+
+const fs = require('fs');
+const path = require('path');
+
+const getMergedConfig = require('../utils/getMergedConfig');
+const getTaglibInvocations = require('./getTaglibInvocations');
+
+const TAGLIBS = {
+	'aui:script': {
+		attributes: ['require'],
+	},
+	'liferay-frontend:component': {
+		attributes: ['module'],
+	},
+	'react:component': {
+		attributes: ['module', 'propsTransformer'],
+	},
+	'soy:template-renderer': {
+		attributes: ['module'],
+	},
+};
+
+/**
+ * Scans JSP files for JS modules and reports linting errors for those not
+ * included in the `exposes` section for the `npmscripts.config.js` file.
+ */
+module.exports = function (source) {
+	const config = getMergedConfig('npmscripts');
+
+	if (!fs.existsSync(config.build.input)) {
+		return [];
+	}
+
+	const invocations = getTaglibInvocations(source, ...Object.keys(TAGLIBS));
+
+	const moduleNames = new Set();
+
+	for (const invocation of invocations) {
+		const {attributes: validAttributes} = TAGLIBS[invocation.tag];
+		const {attributes} = invocation;
+
+		for (const validAttribute of validAttributes) {
+			const require = attributes[validAttribute];
+
+			if (!require) {
+				continue;
+			}
+
+			require
+				.split(',')
+				.map((part) => part.split(' as ')[0])
+				.forEach((moduleName) => moduleNames.add(moduleName));
+		}
+	}
+
+	const messages = [];
+
+	const federation =
+		typeof config.federation === 'object' ? config.federation : {};
+	const exposes = federation.exposes || [];
+
+	for (const moduleName of moduleNames) {
+		if (moduleName.includes('<%=')) {
+			continue;
+		}
+
+		const modulePath = `${moduleName}.js`;
+
+		if (!fs.existsSync(path.join(config.build.input, modulePath))) {
+			continue;
+		}
+
+		if (!exposes.includes(`<inputDir>/${modulePath}`)) {
+			messages.push({
```

Comment:
I'm mimicking eslint messages here because our linter assumes that everything is linted by eslint.

We can probably turn the check-exposes code into an eslint plugin, but I have no experience with them so I wanted to first settle what we have and then we may enhance it if we think it is necessary.

However, I don't know what others may think... Feedback welcome.

Reply:
> We can probably turn the check-exposes code into an eslint plugin, but I have no experience with them so I wanted to first settle what we have and then we may enhance it if we think it is necessary.

I think mimicking the message format is fine. You don't want a real ESLint plugin here because you're not linting ES.

(That said, I haven't read the code yet. I'm about to go and do that now.)